### DISCUSSION TITLE
Find link index

### DISCUSSION THREAD

**Discussion by sven1904** [bulletphysics/bullet3#4026] [Q&A]
Am I right that there's no way to programmatically find out the link index by the link name used in the URDF?
Something like `pybullet.getLinkIndex('urdf_link_name')` or vice versa `pybullet.getLinkName(1)`

Background:
I'm trying to setup some contact constraints between links, but I don't have a clue how bullet introduces link indices since they do not always match the order in the URDF file.
For example, if you load [this model](https://github.com/osrf/gazebo_tutorials/raw/master/kinematic_loop/four_bar_split_fixed_urdf/model.urdf), the resulting link index->name map is this:
```
-1 -> world
 0 -> link_AB
 1 -> link_BE
 2 -> link_CD
 3 -> link_EC
```
although the URDF order is different for the last two elements and the [URDFParser](https://github.com/bulletphysics/bullet3/blob/98254b51a934371e6c6bab6e42d98784c79b3598/examples/Importers/ImportURDFDemo/UrdfParser.cpp#L2117) goes through the file link by link.

Any information would be appreciated.

PS: I know that the gazebo tags are ignored in the URDF file, it is only an example file from the web.

**Comment by rohit-kumar-j** [comment]
@sven1904 you can either create a utility function or a utility class that does that programmatically when the robot is loaded in or compare the `p.getJointInfo[i]` values (where 'i' is the index of the name of the joint). 

....Something like that as far as I remember

**Comment by sven1904** [comment]
hello @rohit-kumar-j,

thanks for your suggestion.

yes, for the joints there is the method `getJointInfo(i)` where you also get the information on the joint's name.
unfortunately, there is no such function for the links, but i need the link indices for the creation of constraints.
even if i go with the workaround via `getJointInfo(i)` which also gives me the **index** of the parent link and the **name** (why?) of the child, i still don't know the index of the child.

maybe a short example for explanation:
i have a model of a robot hand with five fingers and want to add a contact constraint to the pointer finger tip and the tip of the thumb.
with `getJointInfo(i)`, i can find out all indices of finger links except for the tip.
so, there are still 5 possibilities left for each of the tips.
a workaround for this would be to add another "virtual" joint and link to each of the fingers and get the parent index of that joint via a loop over `getJointInfo(i)` with joint name comparison.
but this seems quite complex for such a simple information and messing around with the model is not that nice for that kind of stuff.

this is why i asked if there is already a functionality (without editing the model) to find out which link index refers to which link name.

**Comment by sven1904** [comment]
after having a look [here](https://github.com/bulletphysics/bullet3/blob/fa50f195e6911d60c5ed888d5df33adb1736f93b/examples/pybullet/gym/pybullet_envs/minitaur/envs/minitaur.py#L378), i'm wondering if the child link always has the same index as the joint to which it is appended.
because although `createConstraint(..)` takes link indices as parameters, there are joint indices given.

@erwincoumans as the author, can you confirm that this is the case?

**Comment by rohit-kumar-j** [comment]
@sven1904 
Yes, it is cumbersome to find the link indices manually, however in my experience when you load the URDF, even if you do not mention the `p.URDF_MAINTAIN_LINK_ORDER` argument to the `p.loadURDF()` function, the links spawn with the same indices.

I simply color the links with `p.changeVisualShape()` to input the link that I want to color, I think the links are +or- 1 to the joint index. That way I can find the link Index. It is a little troublesome, but only once per URDF.

I had created constraints here: https://youtu.be/0hfkm7vv5o8

Also See: #3924
